### ISSUE TITLE
Env.runLayer ignores pending effects introduced by layers

### ISSUE THREAD

**Issue by johnhungerford** [getkyo/kyo#603]
Failing test:

```scala3
            "should provide all layers" in {
                val effect: Int < Env[String & Int & Boolean & Char] =
                    Env.get[String] *> Env.get[Int] *> Env.get[Boolean] *> Env.get[Char].as(23)
                val layerChar   = Layer(IO('c'))
                val layerString = Layer("value")
                val layerInt    = Layer(1)
                val layerBool   = Layer(false)
                val handled =
                    Env.runLayer(
                        layerChar,
                        layerString,
                        layerInt,
                        layerBool
                    )(effect)
                assert(Memo.run(handled).eval == 23)
            }
```

The above should fail with a compilation error. `handled` should have type `Int < (IO & Memo)` because `layerChar` has an `IO` effect input. Instead, `Env.runLayer` casts it to `Int < Memo`. As a result, it compiles but then fails at runtime with `kyo.bug$KyoBugException: Unexpected pending effect`

**Comment by johnhungerford** [comment]
I think I have a fix. Here is what `Env.runLayer` should look like:

```scala3
    transparent inline def runLayer[A, S, V](inline layers: Layer[?, ?]*)(value: A < (Env[V] & S)): A < (S & Memo & Nothing) =
        inline Layer.init[V](layers*) match
            case layer: Layer[V, s] =>
                layer.run.map { env =>
                    runTypeMap(env)(value)
                }
```

**Comment by johnhungerford** [comment]
Have a PR: #604 -- I'm seeing if I can get a `provideSome` method working too...

**Comment by fwbrasil** [comment]
nice! that was quick :)

**Comment by hearnadam** [comment]
Fixed by #604
